### PR TITLE
RFC - coverage declaration

### PR DESCRIPTION
**Pull request by cbeauchesne**

### CODE REVIEW

**Comment by cbeauchesne** [comment]
Was a missclick ...

File: docs/internals/RFC-feature-coverage.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+The test fully covers the core functionnality of the feature. The test will catch all bug that can happen, except for edge cases and extreme scenarios
+
+### Complete test
+
+The test fully cover the functionnality, and missing a bug is statistically negligible => If a bug is missed, we should be included in the post mortem and review our understanding of the feature
+
+----
+
+There is three level for an implemented test. It's a compromise between too many levels with blurry boundaries, and not enough to describe what we want. Fell free to challenge them.
```
Comment (simon-id) [inline docs/internals/RFC-feature-coverage.md:69]:
```suggestion
There is three level for an implemented test. It's a compromise between too many levels with blurry boundaries, and not enough to describe what we want. Feel free to challenge them.
```

**Review by vdeturckheim — APPROVED** [review]
sgtm

File: docs/internals/RFC-feature-coverage.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+The test is the most simple test we can imagine for a feature. It ensure that the feature is working, but does not cover any subtility or edge cases
+
+### Good test
+
+The test fully covers the core functionnality of the feature. The test will catch all bug that can happen, except for edge cases and extreme scenarios
+
+### Complete test
+
+The test fully cover the functionnality, and missing a bug is statistically negligible => If a bug is missed, we should be included in the post mortem and review our understanding of the feature
```
[the remark this answers is not in the document]
Comment (cbeauchesne) [reply docs/internals/RFC-feature-coverage.md:65]: Very interesting subject :) If by "complete" we mean "perfect", you're perfectly true. A test will never be perfect.

What I mean here is more "We are very confident that a bug cannot happen on this feature if the test passes." Maybe `complete` is not the best word ? (ping @lloeki who loves finding good naming ;) )

[the remark this answers is not in the document]
Comment (lloeki) [reply docs/internals/RFC-feature-coverage.md:65]: I was just thinking about that @simon-id, while trying to understand the exact meaning behind these names and descriptions! ;)

Indeed I think naming this way is definitively very fuzzy, which makes it open to interpretation on both ends (test writer and test result reader). What threshold makes it not "basic" anymore? What is "good": good enough or close to complete? Can a test be reasonably marked as "complete", what if we find a new bug or class of bugs, should "complete" tests that passed with the bug stop being "complete" until the corresponding case is implemented?

@cbeauchesne I think you introduced very interesting, unambiguous descriptions right above, in ways that are clear-cut and barely leave room for interpretation, thus carrying good semantic value:

- `indirect`: the test works by inferring a result from state/globals/behaviour that is not related to the feature itself, i.e using inductive logic. consequence: unrelated moving parts may break the test, but having it pass is a good signal that things are probably working as expected
- `smoke`: the test is extremely minimal and essentially checks whether hitting the feature does not produce breakage: it doesn't test that the feature works, it tests that the feature isn't 
- `minimal`: smoke test + tests that the feature actually works, in the smallest possible sense, e.g happiest path, smallest non-thorough value check. rule of thumb: if you remove anything it doesn't test "feature is working" anymore and thus reverts back to be a smoke test; hence "minimal".
- `happy`: Attempts to test multiple happy paths beyond minimal
- `unhappy`: Attempts to test one or more unhappy paths
- `corners`: Attempts to test identified corner cases, this qualifier does not reflect thoroughness of testing each possible corner, it reflects the difference between testing the most common happy+unhappy paths vs going beyond, with specific, trickier cases.
- `thorough`: Happy and unhappy paths are considered broadly tested. Might be an attribute instead of a decorator, e.g `happy(thorough: true)`, `corners(thorough: true)`...

These could be combinable, e.g:
- minimal+happy+unhappy: both happy and unhappy are minimally tested in a non-inferred way
- happy(thorough)+corners: many positive tests, no negative tests, a few corner cases
- indirect+smoke: smoke tested via inference

I feel this way we can eyeball testing coverage with confidence, possibly even compute a synthetic per test+total score from it.

File: docs/internals/RFC-feature-coverage.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+```
+
+Or
+
+```python
+class Test_Feature(BaseTestCase.GoodCoverage):
+    pass
+```
+
+There are pros and cons (see below), which one do you prefer ?
```
Comment (stombre) [inline docs/internals/RFC-feature-coverage.md:22]: Decorators seems better (to me)

File: docs/internals/RFC-feature-coverage.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    pass
+
+# => the coverage is "basic", and the test will be flagged as missing feature for golang/gin
+```
+
+The `@overage` decorator can't be used on methods
+
+### Using base test class inheritance
+
+* Pro: save one code line
```
Comment (stombre) [inline docs/internals/RFC-feature-coverage.md:127]: Not a real pro, you do not paid line of code

File: docs/internals/RFC-feature-coverage.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (stombre) [reply docs/internals/RFC-feature-coverage.md:65]: Only my opinions:
Loic has something, maybe too complex (I fear that we did not fill well the metadata in the long run, if there are too many of them). On the other hand, I think Loic's proposal is relevant. We should think about the complexity of usage vs quality of insights.
If we are ready to pay this complexity, I would support Loic model. Every value could be a decorator.

@lloeki I learn something today (I was not aware of this terminology happy/unhappy path). Thanks for sharing.

[the remark this answers is not in the document]
Comment (cbeauchesne) [reply docs/internals/RFC-feature-coverage.md:65]: Actually, for functional coverage, I think that we'll never be able to get rid of implicit thresholds : we'll never have 100% complete RFC describing every possible use cases, and giving for each of them a notation, so we'll always need to delegate to the test which use case of a feature is a corner cas, or a happy path, etc. And even with this details, you still have an implicit threshold : *are considered broadly tested*.

I totally agree that those thresholds are the pitfall of this feature. But as we cannot eliminate them, my strategy is to limit them to a level of details compatible with how often developers write tests.

One key thing to keep in mind is that they write or modify system tests not so often. Maybe once per month for AppSec engineers. And giving this notation will be done only for those who wrote the test, not when you only enable it.

So having too many detail level will cause a lot of friction, and at the end will give exactly the opposite expected result : no notation or wrongly interpreted notation.

If one day we have a dedicated team to write test, I will agree that more details will be far better, as this team will use them daily. But for now, I think that keeping it simple is very important. I even hesitated with only two levels, using your terms, would be `minimal` and `not-minimal` 😄

File: docs/internals/RFC-feature-coverage.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
+## TL;DR.
+
+We'll implement a mechanism that will put metadata on each test class about the coverage/completness of the test. Two questions : 
+
+* There will be three level, `basic`, `good`, `complete`. It's a compromise, do you think it's too many? not enough?
```
Comment (colin-higgins) [inline docs/internals/RFC-feature-coverage.md:5]: We could consider these as enums for something like a score? What if we allowed a max score of 100 and did something like allow a manual specification of the score if these are not precise enough?
`Basic = 30, Good = 60, Complete = 100`

That said, I think this will be granular enough for most cases anyways.

**Comment by colin-higgins** [comment]
Should the decorators allow setting languages to override as well? Or are these decorators specifically about the tests themselves?

**Comment by cbeauchesne** [comment]
> Should the decorators allow setting languages to override as well? Or are these decorators specifically about the tests themselves?

No, there are strictly about tests.
